src/token/mod.rs

```diff
@@ -13,522 +13,639 @@
         )
         .build()
         .to_vec()
-        .map_err(AccessTokenError::from_cose_error)
+        .map_err(AccessTokenError::from)
+}
+
+/// Encrypts the given `claims` with the given headers and `external_aad` for each recipient
+/// by using the `keys` with the cipher given by type parameter `T`,
+/// returning the token as a serialized bytestring of the [`CoseEncrypt`] structure.
+///
+/// Note that the given `keys` must each have an associated `kid` (key ID) field when converted
+/// to COSE keys, as the recipients inside the [`CoseEncrypt`] are identified in this way.
+///
+/// The Content Encryption Key (used to encrypt the actual claims) is randomly generated by the
+/// given cipher in `T`, whereas the given `keys` are used as Key Encryption Keys, that is,
+/// they encrypt the Content Encryption Key for each recipient.
+///
+/// # Errors
+/// - When there's a [`CoseError`](coset::CoseError) while serializing the given `claims` to CBOR.
+/// - When there's a [`CoseError`](coset::CoseError) while serializing the [`CoseEncrypt`] structure.
+/// - When the given headers conflict with the headers set by the cipher `T`.
+///
+/// # Example
+/// For example, assuming we have a [`MultipleEncryptCipher`] in `FakeCrypto`, a random number
+/// generator in `rng`, and some `claims`, we can then create a token encrypted for two recipients
+/// (with keys `key1` and `key2`, respectively) as follows:
+/// ```ignore
+/// let encrypted = encrypt_access_token_multiple::<FakeCrypto, FakeRng>(
+///    vec![&key1, &key2], claims.clone(), None, None, None rng
+/// )?;
+/// ```
+pub fn encrypt_access_token_multiple<T, RNG>(
+    keys: Vec<&T::EncryptKey>,
+    claims: ClaimsSet,
+    external_aad: Option<&[u8]>,
+    unprotected_header: Option<Header>,
+    protected_header: Option<Header>,
+    mut rng: RNG,
+) -> Result<ByteString, AccessTokenError<T::Error>>
+where
+    T: MultipleEncryptCipher,
+    RNG: CryptoRng + RngCore,
+{
+    let key = T::generate_cek(&mut rng);
+    let (unprotected, protected) =
+        prepare_headers!(&key, unprotected_header, protected_header, &mut rng, T)?;
+    let mut builder = CoseEncryptBuilder::new()
+        .unprotected(unprotected.clone())
+        .protected(protected.clone())
+        .create_ciphertext(
+            &claims.to_vec()?[..],
+            external_aad.unwrap_or(&[0; 0]),
+            |payload, aad| T::encrypt(&key, payload, aad, &protected, &unprotected),
+        );
+    let serialized_key: Vec<u8> = key.into();
+    for rec_key in keys {
+        let (rec_unprotected, rec_protected) = prepare_headers!(rec_key, None, None, &mut rng, T)?;
+        builder = builder.add_recipient(
+            CoseRecipientBuilder::new()
+                .protected(rec_protected.clone())
+                .unprotected(rec_unprotected.clone())
+                .create_ciphertext(
+                    // TODO: What should AAD be here?
```

Comment:
AAD is probably not applicable here.

Reply:
Well, I do need to pass something in, though. Is the current approach (an empty bytearray) fine?

Reply:
Yes, passing in no additional data should be fine.